designs/aws/custom-user-data-and-amis.md

```diff
@@ -0,0 +1,285 @@
+    *
+    ```
+      spec:
+          binpacking:
+            kubeReserved:
+              memory:
+              cpu:
+            systemReserved:
+              memory:
+              cpu:
+    or...
+
+      spec:
+        kubeletConfiguration:
+          kubeReserved:
+            memory:
+            cpu:
+   ```
+
+* The main advantage of this approach is that we don’t need to try and determine what kind of OS that AMI has and then try and predict what kind of UserData might work. That’s extremely brittle.
+* Users are already doing this today in Karpenter (and in all nodegroups options too) and while there have been some sharp edges [see this](https://github.com/aws/karpenter/issues/1380), for the most part this does work and is simplest to reason about.
+
+
+_Option 2 - Let users indicate what bootstrapping logic to use by defining AMI class_
+
+In this option, users tell Karpenter to use whatever bootstrapping logic they’d normally use for other AMIFamilies. Our APIs would evolve as -
+
+
+spec:
+  provider:
+    apiVersion: extensions.karpenter.sh/v1alpha1
+    kind: AWS
+    ami: ami-123456
+    amiFamily: AL2
+    userData: "ba123bc.." #base64 encoded
+
+* In such cases, Karpenter would operate the exact same way it does when the amiFamily is AL2. It will attempt to bootstrap the worker node as if it were an EKS optimized AL2 AMI.
+* This is helpful for users that build their AMIs using the packer scripts we’ve got (https://github.com/awslabs/amazon-eks-ami) and only perform minimal modifications.
+* This will not cover all use cases. It’s difficult for us to know the extent of customizations that users are currently doing, but we’re probably going to be excluding a set of use cases.
+* We’d have to be extremely careful modifying how Karpenter bootstraps worker nodes. While our new logic might work for new EKS AMIs, we may end up breaking custom AMIs.
+
+------
+------
+
+### Supporting other LaunchTemplate Parameters
+
+We’ve heard of other instance parameters that users want to control, which Karpenter does not need any control over and can wire through completely.
+
+1. [Capacity Reservations](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_LaunchTemplateCapacityReservationSpecificationRequest.html) - RIs for higher instance availabilities where you’d want to wire these values down to the EC2 Fleet request.
+2. [Cpu Options](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_LaunchTemplateCpuOptionsRequest.html) - If you wanted to control the ratio of mem / cpu on every instance type that you get back. This can be useful if you want to force a certain mem / cpu ratio per instance.
+
+We could potentially also add `ElasticInferenceAccelerators`, `ElasticGPUSpecifications` and some others if a use case arises.
+
+------
+------
+
```

Comment:
There are a few double empty lines in the file, not sure if that serves any purpose.

Reply:
I don't think the formatting is something I'm really concerned with addressing before merging in. The double lines were just things I'd added in as line separators. 

Apologies for some of the typos, I'll get those fixed.